pandas/core/dtypes/concat.py

```diff
@@ -95,17 +95,23 @@ def is_nonempty(x) -> bool:
     _contains_datetime = any(typ.startswith("datetime") for typ in typs)
     _contains_period = any(typ.startswith("period") for typ in typs)
 
+    from pandas.core.arrays import Categorical, SparseArray, datetimelike as dtl
```

Comment:
And to make my suggestion more concrete: instead of import Categorical here, it would be `from pandas.core.arrays.categorical import _concat_arrays as concat_categorical` (or whathever name we give it)